### ISSUE TITLE
parameter setting seems to be wrong

### ISSUE THREAD

**Issue by JackAILab** [kiva12138/MITRL#4]
Dear author

I find that the a_dim, v_dim, and t_dim may be incorrect, beacuase the self.d_k % n_head should be zero in your Transformer block.

How to slove this?
![image](https://user-images.githubusercontent.com/112920087/233609898-6419072e-70fc-4c1f-aee0-759832bf971e.png)

When I change these three parameters, another BUG appear. So how can I correctly solve this?
![image](https://user-images.githubusercontent.com/112920087/233610131-1eb80143-ccf5-4dca-bf16-06bb7eb948ca.png)

**Comment by kiva12138** [comment]
Actually, a_dim, t_dim, and v_dim % n_head don't have to be 0. But d_k % n_head should be 0. 

It seems that you have employed your own data. The input shape should be [batch_size, temporal_length, feature_channel]. Is your data with this shape?

**Comment by JackAILab** [comment]
> 实际上，a_dim、t_dim 和 v_dim % n_head 不必为 0。但 d_k % n_head 应该为 0。
> 
> 您似乎使用了自己的数据。输入形状应该是 [batch_size, temporal_length, feature_channel]。你的数据是这种形状吗？

Oh, I have solved this BUG, just set d_k=512, d_v=1024, no need to change other parameters. But I have another question, how much GPU computing is actually used in your project? I used a 24G 3090 and it couldn't run. I set the batchsize to 1 and set it to run the model with a single card.

<img width="1834" alt="1682151385345" src="https://user-images.githubusercontent.com/112920087/233772117-83a099b8-9a7b-4c77-ba74-053af981dfb4.png">

**Comment by JackAILab** [comment]
> 实际上，a_dim、t_dim 和 v_dim % n_head 不一定为 0。但 d_k % n_head 应为 0。
> 
> 你好像使用了自己的数据。输入形状应该是 [batch_size, temporal_length, feature_channel]。你的数据是这种形状吗？

哦，我已经解决这个BUG了，只需要设置d_k=512, d_v=1024,即可，无需更改其他参数。但是我还有一个问题，你的项目中实际是使用多少GPU运算的？我使用一张24G的3090竟然无法运行，我将batchsize设置为1，并且设置为了单卡运行模型。

**Comment by JackAILab** [comment]
> 实际上，a_dim、t_dim 和 v_dim % n_head 不一定为 0。但 d_k % n_head 应为 0。
> 
> 你好象使用了自己的数据。输入形状应该是[batch_size, temporal_length, feature_channel]。你的数据是这种形状吗？

Oh, I have solved this BUG, just set d_k=512, d_v=1024, no need to change other parameters. But I have another question, how much GPU computing is actually used in your project? I used one 24G 3090 and it couldn't run. I set the batchsize to 1 and set it to run the model with a single card.

**Comment by JackAILab** [comment]
> 实际上，a_dim、t_dim 和 v_dim % n_head 不一定为 0。但 d_k % n_head 应为 0。
> 
> 你好象使用了自己的数据。输入形状应该是[batch_size, temporal_length, feature_channel]。你的数据是这种形状吗？

![image](https://user-images.githubusercontent.com/112920087/233772760-4416319a-0a45-4a66-b439-cf3945d2b68b.png)

**Comment by kiva12138** [comment]
We use two 1080ti GPU cards and set batch size to 64 in our experiments. But the dimensions of the data in our employed dataset are less than 300. I think you can reduced the sequential length if your memory is out, e.g., word-level alignment.

**Comment by JackAILab** [comment]
> We use two 1080ti GPU cards and set batch size to 64 in our experiments. But the dimensions of the data in our employed dataset are less than 300. I think you can reduced the sequential length if your memory is out, e.g., word-level alignment.

You mean the parameter of t_dim? I set these parameters in default, and I find that the gpu demand is too high, maybe I set something wrong?
![image](https://user-images.githubusercontent.com/112920087/233777967-d9402bb8-2d6a-4a59-8452-3653eabb14a7.png)

**Comment by kiva12138** [comment]
--layer means the number of layers in the model. 2048 layers is too high, try to set --layer less than 5.

**Comment by JackAILab** [comment]
> --layer means the number of layers in the model. 2048 layers is too high, try to set --layer less than 5.

OK, that seems to be effective. However, the parameter class maybe wrong? because the label got one BUG as shown in the following screenshot
![image](https://user-images.githubusercontent.com/112920087/233782352-58ce8d01-0016-4f91-9d1c-5644f0653d8f.png)

![image](https://user-images.githubusercontent.com/112920087/233782551-72ab54d0-9a9c-4247-ad61-5c62cb24858e.png)

**Comment by JackAILab** [comment]
> > --layer means the number of layers in the model. 2048 layers is too high, try to set --layer less than 5.
> 
> OK, that seems to be effective. However, the parameter class maybe wrong? because the label got one BUG as shown in the following screenshot ![image](https://user-images.githubusercontent.com/112920087/233782352-58ce8d01-0016-4f91-9d1c-5644f0653d8f.png)
> 
> ![image](https://user-images.githubusercontent.com/112920087/233782551-72ab54d0-9a9c-4247-ad61-5c62cb24858e.png)

I also change other loss parameters, BUG still here
![image](https://user-images.githubusercontent.com/112920087/233782843-42750741-58e0-4556-a12d-1cbfa768b9c6.png)

**Comment by kiva12138** [comment]
It depends on your dataset and tasks. For 3-class classification, num_class should be set to 3.  --num_class==7 means the target task is 7-class classification.

**Comment by JackAILab** [comment]
> It depends on your dataset and tasks. For 3-class classification, num_class should be set to 3. --num_class==7 means the target task is 7-class classification.

OK, I got it, I can run all these codes now. It's maybe one last BUG, when I use the the MOSI_20 dataset, the make_weights function can not well handle the multi-label, because the label of this MODI_20 dataset contain three labels. So I just use the last label as shown in the following screenshot, am I right for doing that?

Besides, I am not sure that if the dataset MOSI_20 is the same with the dataset MOSI, because I can't find the MOSI_20 dataset in your provided Google Drive. So I just copy this dataset into MOSI another dataset fold named MOSI_20. Am I right for ding so? The same problem with the dataset of IEMOCAP_20, MOSI_50, MOSEI_20, MOSEI_50.

![image](https://user-images.githubusercontent.com/112920087/233788778-905ea1c7-2ea7-45b1-9084-ba1943dc64ae.png)

![image](https://user-images.githubusercontent.com/112920087/233789200-2f4bc25b-56e4-44b6-b91c-3902d6982258.png)

**Comment by JackAILab** [comment]
> > 这取决于您的数据集和任务。对于3级分类，num_class应该设置为3。--num_class==7表示目标任务是7级分类。
> 
> 好的，我明白了，我现在可以运行所有这些代码了。这可能是最后一个BUG，当我使用MOSI_20数据集时，make_weights函数不能很好地处理多标签，因为这个MODI_20数据集的标签包含三个标签。所以我只使用最后一个标签，如下面的屏幕截图所示，我这样做是否正确？
> 
> 此外，我不确定数据集 MOSI_20 是否与数据集 MOSI 相同，因为我在您提供的 Google Drive 中找不到 MOSI_20 数据集。所以我只是将这个数据集复制到 MOSI 另一个名为 MOSI_20 的数据集折叠中。我适合吗？IEMOCAP_20、MOSI_50、MOSEI_20、MOSEI_50 的数据集也存在同样的问题。
> 
> ![图像](https://user-images.githubusercontent.com/112920087/233788778-905ea1c7-2ea7-45b1-9084-ba1943dc64ae.png)
> 
> ![图像](https://user-images.githubusercontent.com/112920087/233789200-2f4bc25b-56e4-44b6-b91c-3902d6982258.png)

**Comment by JackAILab** [comment]
> > > 这取决于您的数据集和任务。对于3级分类，num_class应该设置为3。--num_class==7表示目标任务是7级分类。
> > 
> > 
> > 好的，我明白了，我现在可以运行所有这些代码了。这可能是最后一个BUG，当我使用MOSI_20数据集时，make_weights函数不能很好地处理多标签，因为这个MODI_20数据集的标签包含三个标签。所以我只使用最后一个标签，如下面的屏幕截图所示，我这样做是否正确？
> > 此外，我不确定数据集 MOSI_20 是否与数据集 MOSI 相同，因为我在您提供的 Google Drive 中找不到 MOSI_20 数据集。所以我只是将这个数据集复制到 MOSI 另一个名为 MOSI_20 的数据集折叠中。我适合吗？IEMOCAP_20、MOSI_50、MOSEI_20、MOSEI_50 的数据集也存在同样的问题。
> > ![图像](https://user-images.githubusercontent.com/112920087/233788778-905ea1c7-2ea7-45b1-9084-ba1943dc64ae.png)
> > ![图像](https://user-images.githubusercontent.com/112920087/233789200-2f4bc25b-56e4-44b6-b91c-3902d6982258.png)

Unfortunately, I can't reproduce the results in your paper. So some wrong parameter setting still here? The prediction results seem to be bad.
![image](https://user-images.githubusercontent.com/112920087/233790369-822cb78a-683b-4a84-869f-1eb0ac5f90e3.png)

**Comment by kiva12138** [comment]
The performance can be affected by many issues, e.g., CUDA version and pytorch version. I think you can tune some hyperparameters to make it better, e.g., learning rate and loss coefficients.
